### PR TITLE
librbd: differentiate journal replay flush vs shut down

### PR DESCRIPTION
**Pull request by dillaman**
Signed-off-by: Jason Dillaman <dillaman@redhat.com>

### CODE REVIEW

**Comment by dillaman** [comment]
@trociny This should help your test cases

**Comment by trociny** [comment]
@dillaman it helps with the crushes I observed when using Replay::flush() before. Still a minor issue: the replayer client commit position is not updated immediately (before the flush command returns). In my tests I have to wait about 5-10 sec to for the replayer client position to match the master one.

**Comment by dillaman** [comment]
The Replay class doesn't have access to the journal::Journaler, so it cannot flush the commit position.  I think I would need to add a flush method to librbd::Journal to accomplish that goal.

**Comment by dillaman** [comment]
@trociny I added a new ```librbd::Journal::flush_commit_position``` helper method that you can invoke after getting a completion from ```librbd::journal::Replay::flush```.  Let me know if you have any issues with it.

File: src/journal/JournalMetadata.h [hunk 1]
```diff
@@ -159,6 +160,24 @@ class JournalMetadata : public RefCountedObject, boost::noncopyable {
     }
   };
 
+  struct C_FlushCommitPosition : public Context {
+    JournalMetadata *journal_metadata;
```
Comment (jdurgin) [inline src/journal/JournalMetadata.h]: looks like this is unused

**Comment by dillaman** [comment]
@jdurgin I was planning to track the op, but it's wrapped by ```C_NotifyUpdate``` which already handles that scenario.

**Comment by trociny** [comment]
@dillaman flush_commit_position makes commit position update much faster, thanks!

Still it looks a little strange to me, that I need to run the flush several times to update the position to the latest. Here is my current ImageReplayer::flush() implementation:

https://github.com/trociny/ceph/blob/wip-rbd-mirroring-image-reply-wip-librbd-replay-locks/src/tools/rbd_mirror/ImageReplayer.cc#L307

And here is the test:

https://github.com/trociny/ceph/blob/wip-rbd-mirroring-image-reply-wip-librbd-replay-locks/src/test/rbd_mirror/test_ImageReplayer.cc#L334

The test output:

```
[ RUN      ] TestImageReplayer.WriteAndStartReplay
m_replayer->flush()
get_commit_positions: client: [id=, commit_position=[object_number=1, positions=[[tag_tid=0, entry_tid=1]]]]
get_commit_positions: client: [id=7933fdd1-05a8-4429-bb9a-c959d82a322a, commit_position=[object_number=0, positions=[]]]
m_replayer->flush()
get_commit_positions: client: [id=, commit_position=[object_number=1, positions=[[tag_tid=0, entry_tid=1]]]]
get_commit_positions: client: [id=7933fdd1-05a8-4429-bb9a-c959d82a322a, commit_position=[object_number=1, positions=[[tag_tid=0, entry_tid=1]]]]
wrote: 512
wrote: 512
wrote: 512
wrote: 512
wrote: 512
wrote: 512
wrote: 512
wrote: 512
wrote: 512
wrote: 512
flushed
m_replayer->flush()
get_commit_positions: client: [id=, commit_position=[object_number=0, positions=[[tag_tid=0, entry_tid=12]]]]
get_commit_positions: client: [id=7933fdd1-05a8-4429-bb9a-c959d82a322a, commit_position=[object_number=2, positions=[[tag_tid=0, entry_tid=2]]]]
m_replayer->flush()
get_commit_positions: client: [id=, commit_position=[object_number=0, positions=[[tag_tid=0, entry_tid=12]]]]
get_commit_positions: client: [id=7933fdd1-05a8-4429-bb9a-c959d82a322a, commit_position=[object_number=1, positions=[[tag_tid=0, entry_tid=5]]]]
m_replayer->flush()
get_commit_positions: client: [id=, commit_position=[object_number=0, positions=[[tag_tid=0, entry_tid=12]]]]
get_commit_positions: client: [id=7933fdd1-05a8-4429-bb9a-c959d82a322a, commit_position=[object_number=0, positions=[[tag_tid=0, entry_tid=8]]]]
m_replayer->flush()
get_commit_positions: client: [id=, commit_position=[object_number=0, positions=[[tag_tid=0, entry_tid=12]]]]
get_commit_positions: client: [id=7933fdd1-05a8-4429-bb9a-c959d82a322a, commit_position=[object_number=3, positions=[[tag_tid=0, entry_tid=11]]]]
m_replayer->flush()
get_commit_positions: client: [id=, commit_position=[object_number=0, positions=[[tag_tid=0, entry_tid=12]]]]
get_commit_positions: client: [id=7933fdd1-05a8-4429-bb9a-c959d82a322a, commit_position=[object_number=0, positions=[[tag_tid=0, entry_tid=12]]]]
read: 512
read: 512
read: 512
read: 512
read: 512
read: 512
read: 512
read: 512
read: 512
read: 512
[       OK ] TestImageReplayer.WriteAndStartReplay (13039 ms)
```

**Comment by dillaman** [comment]
@trociny This case is only needed to support the unit tests, so it doesn't seem too awkward.

**Pull request merged**